tephra-core/src/main/java/org/apache/tephra/txprune/TransactionPruningRunnable.java

```diff
@@ -0,0 +1,122 @@
+ * under the License.
+ */
+
+package org.apache.tephra.txprune;
+
+import org.apache.tephra.Transaction;
+import org.apache.tephra.TransactionManager;
+import org.apache.tephra.util.TxUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeSet;
+
+/**
+ * This class executes one run of transaction pruning every time it is invoked.
+ * Typically, this class will be scheduled to run periodically.
+ */
+public class TransactionPruningRunnable implements Runnable {
+  private static final Logger LOG = LoggerFactory.getLogger(TransactionPruningRunnable.class);
+
+  private final TransactionManager txManager;
+  private final Map<String, TransactionPruningPlugin> plugins;
+  private final long txMaxLifetimeMillis;
+
+  public TransactionPruningRunnable(TransactionManager txManager, Map<String, TransactionPruningPlugin> plugins,
+                                    long txMaxLifetimeMillis) {
+    this.txManager = txManager;
+    this.plugins = plugins;
+    this.txMaxLifetimeMillis = txMaxLifetimeMillis;
+  }
+
+  @Override
+  public void run() {
+    try {
+      // TODO: TEPHRA-159 Start a read only transaction here
+      Transaction tx = txManager.startShort();
+      txManager.abort(tx);
+
+      long now = System.currentTimeMillis();
+      long inactiveTransactionBound = TxUtils.getPruneUpperBound(tx, txMaxLifetimeMillis, now);
+      LOG.info("Starting invalid prune run for time {} and inactive transaction bound {}",
+               now, inactiveTransactionBound);
+
+      List<Long> pruneUpperBounds = new ArrayList<>();
+      for (Map.Entry<String, TransactionPruningPlugin> entry : plugins.entrySet()) {
+        String name = entry.getKey();
+        TransactionPruningPlugin plugin = entry.getValue();
+        try {
+          LOG.debug("Fetching prune upper bound using plugin {}", name);
+          long pruneUpperBound = plugin.fetchPruneUpperBound(now, inactiveTransactionBound);
+          LOG.debug("Got prune upper bound {} from plugin {}", pruneUpperBound, name);
+          pruneUpperBounds.add(pruneUpperBound);
+        } catch (Exception e) {
+          LOG.error("Aborting invalid prune run for time {} due to exception from plugin {}", now, name, e);
+          return;
+        }
+      }
+
+      long minPruneUpperBound = Collections.min(pruneUpperBounds);
+      LOG.info("Got minimum prune upper bound {} across all plugins", minPruneUpperBound);
+      if (minPruneUpperBound <= 0) {
+        LOG.info("Not pruning invalid list since minimum prune upper bound ({}) is less than 1", minPruneUpperBound);
+        return;
+      }
+
+      long[] invalids = tx.getInvalids();
+      TreeSet<Long> toTruncate = new TreeSet<>();
+      LOG.debug("Invalid list: {}", invalids);
+      for (long invalid : invalids) {
+        if (invalid <= minPruneUpperBound) {
+          toTruncate.add(invalid);
```

Comment:
do we have to compute this list? Or can the txManager simply remove all transactions that are older than the minPruneUpperBound?